### PR TITLE
Added more flexibility for source factors combination

### PR DESCRIPTION
**Pull request by davvil**
Added more flexibility for source fators combination

- The combination strategy, sum, concat or average (new) can be
  specified for each factor individually
- Source factors can share the same embedding matrix as the input

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [ ] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [ ] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/arguments.py [hunk 1]
```diff
@@ -645,8 +672,14 @@ def add_model_parameters(params):
                                    '(validation) source factor files. Default: %(default)s.')
     model_params.add_argument('--source-factors-combine', '-sfc',
                               choices=C.SOURCE_FACTORS_COMBINE_CHOICES,
-                              default=C.SOURCE_FACTORS_COMBINE_CONCAT,
+                              default=[C.SOURCE_FACTORS_COMBINE_CONCAT],
+                              nargs='+',
                               help='How to combine source factors. Default: %(default)s.')
```
Comment (fhieber) [inline sockeye/arguments.py]: could you add a sentence in help to indicate that this is a list?

File: sockeye/arguments.py [hunk 2]
```diff
@@ -645,8 +672,14 @@ def add_model_parameters(params):
...
+    model_params.add_argument('--source-factors-share-embedding',
+                              type=bool_str(),
+                              nargs='+',
+                              default=[False],
+                              help='Share the embeddings with the source language. Default: do not share.')
```
Comment (fhieber) [inline sockeye/arguments.py]: could you add a sentence in help to indicate that this is a list?

File: sockeye/encoder.py [hunk 3]
```diff
@@ -93,10 +93,16 @@ def get_max_seq_len(self) -> Optional[int]:
 
 class FactorConfig(config.Config):
 
-    def __init__(self, vocab_size: int, num_embed: int) -> None:
+    def __init__(self,
+                 vocab_size: int,
+                 num_embed: int,
+                 combine: str, # From C.SOURCE_FACTORS_COMBINE_CHOICES
```
Comment (fhieber) [inline sockeye/encoder.py]: any value in making this an Enum?

File: sockeye/encoder.py [hunk 4]
```diff
@@ -143,20 +147,38 @@ def __init__(self,
...
-                    self.factor_embeds.add(mx.gluon.nn.Embedding(fc.vocab_size, fc.num_embed,
-                                                                 prefix="factor%d_" % i))
-
-    def hybrid_forward(self, F, data, valid_length, embed_weight):  # pylint: disable=arguments-differ
-        factor_embeds = []
+                for i, fc in enumerate(self.config.factor_configs):
+                    factor_weight_name = 'factor%d_weight' % i
+                    factor_weight = embed_weight if fc.share_source_embedding else \
+                        self.params.get('factor%d_weight' % i, shape=(fc.vocab_size, fc.num_embed))
+                    # We set the attribute of the class to trigger the hybrid_forward parameter creation "magic"
```
Comment (fhieber) [inline sockeye/encoder.py]: What is the magic here?

File: sockeye/encoder.py [hunk 5]
```diff
@@ -143,20 +147,38 @@ def __init__(self,
...
+                    setattr(self, factor_weight_name, factor_weight)
+
+    def hybrid_forward(self, F, data, valid_length, embed_weight, **kwargs):  # pylint: disable=arguments-differ
+        # We will catch the optional factor weights in kwargs
```
Comment (fhieber) [inline sockeye/encoder.py]: not necessarily of type symbol, rather `Union[Symbol, ndarray]`

File: sockeye/arguments.py (hunk 1 above)
Comment (davvil) [reply sockeye/arguments.py]: Done, although the help produced by argparse already shows this.

File: sockeye/arguments.py (hunk 2 above)
Comment (davvil) [reply sockeye/arguments.py]: Done, although the help produced by argparse already shows this.

File: sockeye/encoder.py (hunk 3 above)
Comment (davvil) [reply sockeye/encoder.py]: It's a bit cleaner, although the whole parameter validation is already performed by argparse.

File: sockeye/encoder.py (hunk 4 above)
Comment (davvil) [reply sockeye/encoder.py]: mxnet defines the parameters of the block (and thus passed to hybrid_forward) as the members of the block which are of class Parameter. The goal is to use something like
```
self.my_param = self.params.get('blablabla', ...)
```
But in our case we create the parameters dynamically. Simply storing them in a list does not work (`self.factor_weights = [self.params.get(...) for ...]`), as `factor_weights` would be of type List. Therefore we create a number of class attributes of the appropriate class.

**Pull request merged**